
### **Planejamento da Limpeza**

Antes de iniciar a Análise Exploratória de Dados (EDA), será necessário realizar uma etapa de **limpeza dos dados** para preparar a base de forma adequada. Essa fase terá como objetivo organizar e padronizar os dados, evitando que a EDA fique sobrecarregada com tarefas de limpeza e garantindo que a análise seja mais objetiva.

As ações previstas para essa etapa são:

1. **Remoção de colunas inúteis:**
   Irei identificar e remover colunas que não tenham relevância para a análise ou para a modelagem, como identificadores sem valor analítico, informações redundantes ou atributos que não contribuem para o problema de negócio.

2. **Tratamento de outliers:**
   Será realizada a análise e o tratamento dos outliers apenas no conjunto de **treinamento**. Essa escolha é importante, pois o conjunto de teste deve simular um cenário de produção. Portanto, não se deve alterar o teste — dessa forma, o modelo será avaliado em condições mais realistas.

3. **Identificação e tratamento de valores ausentes (missing values):**
   Irei verificar a existência de valores ausentes e definir estratégias de tratamento, que podem incluir:

   * remoção de registros com muitos dados faltantes,
   * imputação de valores por estatísticas como média, mediana ou moda,
   * ou aplicação de métodos específicos dependendo do tipo de variável.

Esse processo de **pré-EDA** garantirá que, ao chegar na análise exploratória, os dados já estejam limpos e padronizados, evitando retrabalho e permitindo que a atenção esteja focada na extração de insights e no entendimento das variáveis.




In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

In [4]:
used_cars = pd.read_csv("../Dados/cars.csv")
used_cars.head()

manufacturer            model  year   mileage  \
0        Acura  ILX Hybrid 1.5L  2013   92945.0   
1        Acura  ILX Hybrid 1.5L  2013   47645.0   
2        Acura  ILX Hybrid 1.5L  2013   53422.0   
3        Acura  ILX Hybrid 1.5L  2013  117598.0   
4        Acura  ILX Hybrid 1.5L  2013  114865.0   

                                              engine   transmission  \
0  1.5L I-4 i-VTEC variable valve control, engine...      Automatic   
1                        1.5L I4 8V MPFI SOHC Hybrid  Automatic CVT   
2                        1.5L I4 8V MPFI SOHC Hybrid  Automatic CVT   
3                        1.5L I4 8V MPFI SOHC Hybrid  Automatic CVT   
4                        1.5L I4 8V MPFI SOHC Hybrid  Automatic CVT   

          drivetrain fuel_type    mpg           exterior_color interior_color  \
0  Front-wheel Drive  Gasoline  39-38                    Black      Parchment   
1  Front-wheel Drive    Hybrid  39-38                     Gray          Ebony   
2  Front-wheel Drive    Hybrid  39-38    Bellanova White Pearl          Ebony   
3  Front-wheel Drive    Hybrid  39-38  Polished Metal Metallic            NaN   
4  Front-wheel Drive    Hybrid  39-38                      NaN          Ebony   

   accidents_or_damage  one_owner  personal_use_only  \
0                  0.0        0.0                0.0   
1                  1.0        1.0                1.0   
2                  0.0        1.0                1.0   
3                  0.0        1.0                1.0   
4                  1.0        0.0                1.0   

                    seller_name  seller_rating  driver_rating  \
0                  Iconic Coach            NaN            4.4   
1                    Kars Today            NaN            4.4   
2  Weiss Toyota of South County            4.3            4.4   
3              Apple Tree Acura            NaN            4.4   
4       Herb Connolly Chevrolet            3.7            4.4   

   driver_reviews_num  price_drop    price  
0                12.0       300.0  13988.0  
1                12.0         NaN  17995.0  
2                12.0       500.0  17000.0  
3                12.0       675.0  14958.0  
4                12.0       300.0  14498.0

In [5]:
print(f"O dataset apresenta {used_cars.shape[0]} linhas e {used_cars.shape[1]} colunas")

O dataset apresenta 762091 linhas e 20 colunas


In [6]:
used_cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 762091 entries, 0 to 762090
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   manufacturer         762091 non-null  object 
 1   model                762091 non-null  object 
 2   year                 762091 non-null  int64  
 3   mileage              761585 non-null  float64
 4   engine               747041 non-null  object 
 5   transmission         752187 non-null  object 
 6   drivetrain           740529 non-null  object 
 7   fuel_type            739164 non-null  object 
 8   mpg                  620020 non-null  object 
 9   exterior_color       753232 non-null  object 
 10  interior_color       705116 non-null  object 
 11  accidents_or_damage  737879 non-null  float64
 12  one_owner            730608 non-null  float64
 13  personal_use_only    737239 non-null  float64
 14  seller_name          753498 non-null  object 
 15  seller_rating    

Todas as colunas apresentam seus valores no tipo certo, agora irei ver se há duplicatas

In [7]:
used_cars.duplicated().sum()

np.int64(9145)

Como ela apresenta essas duplicatas eu irei remove-las para não atrapalha nas modelagens futuras

In [8]:
used_cars.drop_duplicates(inplace=True)
used_cars.duplicated().sum()

np.int64(0)

In [9]:
used_cars.select_dtypes(include="object").describe()

manufacturer      model                      engine       transmission  \
count        752946     752946                      738491             743467   
unique           30      12187                        6903               1313   
top            Ford  Fusion SE  2.0L I4 16V GDI DOHC Turbo  6-Speed Automatic   
freq          78851       3172                       74823             146979   

               drivetrain fuel_type     mpg exterior_color interior_color  \
count              732055    730793  613056         744224         697081   
unique                 33        36     865           7681           4679   
top     Front-wheel Drive  Gasoline   19-26          Black          Black   
freq               238528    637168   15955          59562         283070   

                 seller_name  
count                 744451  
unique                 18254  
top     Gateway Classic Cars  
freq                    1628

In [10]:
used_cars.describe()

year       mileage  accidents_or_damage      one_owner  \
count  752946.000000  7.524680e+05        729288.000000  722176.000000   
mean     2017.790013  5.579103e+04             0.228017       0.562032   
std         5.114479  4.354112e+04             0.419554       0.496137   
min      1915.000000  0.000000e+00             0.000000       0.000000   
25%      2016.000000  2.330500e+04             0.000000       0.000000   
50%      2019.000000  4.561400e+04             0.000000       1.000000   
75%      2021.000000  7.836800e+04             0.000000       1.000000   
max      2024.000000  1.119067e+06             1.000000       1.000000   

       personal_use_only  seller_rating  driver_rating  driver_reviews_num  \
count      728652.000000  541438.000000  721666.000000       752946.000000   
mean            0.657201       4.159330       4.623590           89.692752   
std             0.474645       0.804872       0.276817          115.374787   
min             0.000000       1.000000       1.000000            0.000000   
25%             0.000000       3.800000       4.500000           13.000000   
50%             1.000000       4.500000       4.700000           51.000000   
75%             1.000000       4.700000       4.800000          119.000000   
max             1.000000       5.000000       5.000000         1025.000000   

          price_drop         price  
count  406236.000000  7.529460e+05  
mean     1007.430934  3.520899e+04  
std      1374.759058  1.629960e+06  
min       100.000000  1.000000e+00  
25%       380.000000  1.959500e+04  
50%       643.000000  2.799000e+04  
75%      1007.000000  3.948800e+04  
max    170995.000000  1.000000e+09

Após uma análise inicial da descrição estatística dos dados, o conjunto parece apresentar uma distribuição relativamente coerente, com exceção de um ponto crítico. Na coluna price, foi identificado um valor máximo de 1.000.000.000 (um bilhão de dólares), o que é claramente inviável para um carro, especialmente um veículo usado, caracterizando-se como um outlier extremo.

Para garantir a robustez do modelo, o tratamento deste outlier será realizado seguindo um protocolo rigoroso de prevenção a data leakage. Primeiramente, os dados serão divididos em conjuntos de treino e teste. Todo e qualquer tratamento ou ajuste será aplicado apenas com base nas estatísticas do conjunto de treino, que então serão replicadas no conjunto de teste. Essa abordagem assegura que o modelo não tenha acesso a informações do mundo real (o teste) durante sua fase de treinamento.

É importante salientar que a remoção de outliers será feita com critério. O objetivo não é eliminar todos os veículos de alto valor, pois o mercado inclui carros luxuosos e esportivos com preços elevados, que devem ser representados no modelo. No entanto, o caso específico do valor de um bilhão é um erro evidente de registro de dados e, portanto, será removido para não distorcer completamente o treinamento do algoritmo de regressão.

In [11]:
used_car_train, used_car_test = train_test_split(used_cars, test_size=0.2, random_state=42)

In [12]:
used_car_train = used_car_train[used_car_train["price"]< 1000000000.0]
max(used_car_train["price"])

8888889.0

Após a deparação dos dados e remoção do outlier(só no terino pois no teste só vamos encostar na parte de imputação de valores vazios no pre-processamento) irei análisar os valores vazios e ver quais colunas irei manter ou remover para a análise dos dados.

In [13]:
used_car_train.isna().sum()/used_car_train.shape[0]*100

manufacturer            0.000000
model                   0.000000
year                    0.000000
mileage                 0.062090
engine                  1.910833
transmission            1.256900
drivetrain              2.764151
fuel_type               2.932324
mpg                    18.527612
exterior_color          1.163766
interior_color          7.427514
accidents_or_damage     3.135194
one_owner               4.079488
personal_use_only       3.220360
seller_name             1.132555
seller_rating          28.075968
driver_rating           4.162994
driver_reviews_num      0.000000
price_drop             46.028671
price                   0.000000
dtype: float64

**Justificativa para Remoção da Variável `price_drop`**

A variável `price_drop` será removida do conjunto de dados devido a três problemas fundamentais:

1. **Vazamento de Dados (Data Leakage):**
   - Esta variável é calculada a partir do preço final (`price`), que é nossa variável alvo. Usá-la para prever o preço seria como "trapacear", pois na realidade não sabemos qual será o desconto antes de definir o preço.

2. **Alta Taxa de Valores Nulos:**
   - Cerca de 46,17% dos registros (281522 de 609672) estão faltando nesta coluna. Imputar valores seria artificial e distorceria a análise.

3. **Não é uma Característica do Carro:**
   - O desconto é uma decisão do vendedor, não uma propriedade do veículo. Para agrupar carros por suas características naturais (clusterização) e prever preços com base em seus atributos, esta informação é irrelevante e até enganosa.

**Conclusão:** Remover a coluna é a abordagem mais segura para garantir um modelo válido e generalizável, focado apenas nas características reais dos veículos.

In [14]:
used_car_train.drop(columns=["price_drop"], axis=1, inplace=True)
used_car_test.drop(columns=["price_drop"], axis=1, inplace=True)


#### Situação dos Valores Nulos

O dataset **aparenta apresentar poucos valores nulos** na maioria das colunas, com exceção de:

- **seller_rating**: 28,06% de valores nulos
- **mpg**: 18,61% de valores nulos

Logo irei analisar essas colunas e ver se apresentam alguma correlação com a target, se tiver alguma correlação eu mantnho se não eu só retiro da tabela pra não perder tempo com uma coluna que não irá ter valore nenhum preditivo

In [15]:
used_car_train[["seller_rating", "price"]].corr()

seller_rating    price
seller_rating        1.00000  0.09237
price                0.09237  1.00000

Após a análise realizada, foi possível identificar que essa coluna não apresenta relevância preditiva, portanto não será necessário investir tempo em estratégias de tratamento para ela. Já as demais colunas apresentam uma baixa porcentagem de valores ausentes, não ultrapassando 7%, o que permite a imputação durante o pré-processamento dos modelos. A exceção é a coluna *mgp*, que possui cerca de 18% de dados faltantes. Nesse caso, a imputação simples poderia enviesar os resultados. Além disso, a variável apresenta 829 valores únicos, conforme observado na tabela de descrição das colunas categóricas, o que dificulta a verificação de correlações diretas. Diante disso, será necessário aplicar técnicas de *feature engineering* para tratar essa variável de forma adequada.

In [16]:
used_car_train.drop(columns=["seller_rating"], axis=1, inplace=True)
used_car_test.drop(columns=["seller_rating"], axis=1, inplace=True)

## Feature Engineering

In [17]:
colunas_categoricas = used_car_train.select_dtypes(include="object")
colunas_categoricas.nunique()

manufacturer         30
model             11382
engine             6305
transmission       1163
drivetrain           31
fuel_type            35
mpg                 829
exterior_color     7066
interior_color     4308
seller_name       18105
dtype: int64


O primeiro passo será a remoção de colunas que não trarão utilidade para a análise ou para a modelagem. Conforme observado, o dataset apresenta algumas variáveis com **alta cardinalidade**, ou seja, um número muito elevado de valores únicos, o que dificulta o processamento e pode inviabilizar análises consistentes.

Entre essas colunas estão:

* **`seller_name`**: possui muitos valores distintos, sem relevância preditiva.
* **`model`**: contém 11.381 modelos diferentes, tornando a variável pouco útil e de difícil tratamento.
* **`engine`**: apesar de trazer informações sobre o motor, já existem outras colunas mais detalhadas que cobrem esse aspecto, além de apresentar alta cardinalidade.

Dessa forma, a exclusão dessas colunas permitirá reduzir a dimensionalidade do dataset, eliminar ruído e facilitar a criação de variáveis mais relevantes nas próximas etapas.

In [18]:
used_car_train.drop(columns=["seller_name", "model", "engine"], axis=1, inplace=True)
used_car_test.drop(columns=["seller_name", "model", "engine"], axis=1, inplace=True)

O próximo passo será o tratamento da coluna transmission, responsável por informar o tipo de transmissão do veículo (por exemplo, automática ou manual).

In [19]:
used_car_train["transmission"].value_counts().reset_index().head(50)

transmission   count
0                                   6-Speed Automatic  117453
1                                   8-Speed Automatic  110462
2                                       Automatic CVT   86141
3                                           Automatic   76588
4                                   9-Speed Automatic   48290
5                                  10-Speed Automatic   31632
6                                   5-Speed Automatic   18333
7                                   7-Speed Automatic   15104
8                   7-Speed Automatic with Auto-Shift   12162
9                                      6-Speed Manual   10983
10                                  1-Speed Automatic   10731
11                                  4-Speed Automatic    9576
12                  6-Speed Automatic with Auto-Shift    4884
13                  8-Speed Automatic with Auto-Shift    3981
14                                             Manual    3114
15                                     5-Speed Manual    2799
16                                                A/T    2501
17                 Continuously Variable Transmission    1977
18                                           Variable    1872
19                                                IVT     828
20                                        8-Speed A/T     804
21                                        6-Speed A/T     802
22                                            6-Speed     734
23                                                CVT     703
24                   8-Speed Automatic with Tiptronic     654
25                                 10 Speed Automatic     653
26                     Transmission w/Dual Shift Mode     643
27                  10-Speed Automatic with Overdrive     603
28                                  Automatic 6-Speed     598
29                                  9 Speed Automatic     561
30                                  2-Speed Automatic     513
31                                        6-Speed M/T     452
32                                          AUTOMATIC     437
33                              Automatic Xtronic CVT     427
34                                   Lineartronic CVT     393
35                  8-Speed Automatic with SHIFTRONIC     392
36                                        7-Speed A/T     375
37                          Automatic with Geartronic     365
38                                     7-Speed Manual     335
39                                            8-Speed     329
40                                        5-Speed A/T     301
41                                   CVT Transmission     294
42                                        8-SPEED A/T     292
43        6-Speed Automatic Electronic with Overdrive     278
44                                        9-Speed A/T     253
45                                   CVT with Xtronic     252
46                           ELEC 10-SPEED AUTO TRANS     249
47                                                  A     231
48                            8-SPD AUTO TRANSMISSION     230
49  8 Speed Elect Controlled Auto Transmission w/I...     215

Analisando a amostra acima percebe-se que em sua maioria é automatico ou manual, logo resumirei nessas duas categorias.

In [20]:
def categorizar_transmission(transmission):
    if pd.isna(transmission):
        return np.nan
    
    transmission_lower = str(transmission).lower()
    
    # Manual
    if any(keyword in transmission_lower for keyword in ['manual', 'm/t', 'shift']):
        return 'Manual'
    
    # Automático
    if any(keyword in transmission_lower for keyword in ['automatic', 'a/t', 'cvt', 'variable', 'ivt', 'a', 'elec', 'tiptronic', 'xtronic', 'lineartronic', 'geartronic']):
        return 'Automatic'
    
    # Default para automático (para casos como "6-Speed", "8-Speed" sem especificação)
    return "other"

# Aplicar a categorização
used_car_train['transmission_simple'] = used_car_train['transmission'].apply(categorizar_transmission)
used_car_test['transmission_simple'] = used_car_test['transmission'].apply(categorizar_transmission)
used_car_train.drop(columns=["transmission"], axis=1, inplace=True)
used_car_test.drop(columns=["transmission"], axis=1, inplace=True)
# Verificar a distribuição
transmission_counts = used_car_train['transmission_simple'].value_counts()
print(transmission_counts)

transmission_simple
Automatic    550550
Manual        42410
other          1824
Name: count, dtype: int64


Como visto acima, percebe-se que realmente se resumia em duas categorias mesmo só com algumas diferentes mas numa quantidade muito pequena.

In [21]:
used_car_train.select_dtypes(include="object").nunique()

manufacturer             30
drivetrain               31
fuel_type                35
mpg                     829
exterior_color         7066
interior_color         4308
transmission_simple       3
dtype: int64

Irei analisar agora exterior_color e interior_color.

In [22]:
used_car_train.exterior_color.value_counts().reset_index().head(60)

exterior_color  count
0                                Black  47625
1                                White  40454
2                                 Gray  23981
3                               Silver  19325
4                                 Blue  13759
5                         Summit White  13290
6                                  Red  12117
7               Bright White Clearcoat  10868
8                         Oxford White   6528
9                         Gun Metallic   6192
10  Granite Crystal Clearcoat Metallic   5156
11                         Polar White   4367
12                        Alpine White   4294
13    Billet Silver Metallic Clearcoat   3954
14             Black Sapphire Metallic   3625
15                          Pure White   3601
16                 Silver Ice Metallic   3601
17                          Onyx Black   3539
18                Satin Steel Metallic   3504
19                 Crystal Black Pearl   3303
20                 Crystal White Pearl   3293
21                        Shadow Black   3220
22                       Diamond Black   3191
23                     Black Clearcoat   3101
24                Platinum White Pearl   3090
25                 Ice Silver Metallic   2982
26               Modern Steel Metallic   2939
27              Magnetic Gray Metallic   2862
28           Brilliant Silver Metallic   2861
29                         Super White   2846
30                    Snow White Pearl   2720
31             Midnight Black Metallic   2707
32             Ebony Twilight Metallic   2682
33             Deep Black Pearl Effect   2534
34              Platinum Gray Metallic   2490
35                          Fuji White   2335
36               Lunar Silver Metallic   2312
37                       Glacier White   2284
38                Crystal Black Silica   2235
39               Mosaic Black Metallic   2213
40                           Jet Black   2205
41                Quicksilver Metallic   2140
42                     Santorini Black   2127
43                         Super Black   2079
44                      Magnetic Black   2054
45   Brilliant Black Crystal Pearlcoat   1995
46                Agate Black Metallic   1931
47             Magnetite Gray Metallic   1927
48                         Black Raven   1843
49                               Green   1779
50                      Jet Black Mica   1777
51                   Magnetic Metallic   1774
52                White Frost Tri-Coat   1739
53                      Black Obsidian   1722
54                      Blizzard Pearl   1691
55             White Knuckle Clearcoat   1669
56              Crystal White Tri-Coat   1649
57                Velvet Red Pearlcoat   1603
58              Glacier White Metallic   1592
59                     Graphite Shadow   1554



Com base na amostra observada, foi possível identificar que a maioria das cores se concentra em seis categorias principais: **Black, White, Gray, Silver, Blue** e **Red**. As demais correspondem apenas a variações ou descrições adicionais dessas mesmas cores.

Dessa forma, irei aplicar um **filtro de categorias** para padronização:

* Valores que contenham **“White”** serão agrupados em **White**.
* Valores que contenham **“Black”** serão agrupados em **Black**.
* Valores que contenham **“Gray”** serão agrupados em **Gray**.
* Valores que contenham **“Silver”** ou **“Metallic”** serão agrupados em **Silver**.
* Valores que contenham **“Blue”** serão agrupados em **Blue**.
* Valores que contenham **“Red”** serão agrupados em **Red**.
* Todas as demais variações serão tratadas como pertencentes à categoria correspondente ou agrupadas em uma classe genérica (“Other”), caso não se encaixem em nenhuma das seis principais.

Essa padronização permitirá reduzir a cardinalidade da coluna e facilitar tanto a análise quanto a modelagem.


In [23]:
def categorizar_color_exterior(color):
    if pd.isna(color):
        return np.nan
    
    color_lower = str(color).lower()
    
    # Categorias principais
    if "white" in color_lower:
        return 'white'
    if "black" in color_lower:
        return 'black'
    if "gray" in color_lower or "grey" in color_lower:
        return 'gray'
    if "silver" in color_lower or "metallic" in color_lower:
        return 'silver'
    if "blue" in color_lower:
        return 'blue'
    if "red" in color_lower:
        return 'red'
    
    # Default para cores não mapeadas
    return 'other'


# Aplicar a categorização
used_car_train['color_exterior_simple'] = used_car_train['exterior_color'].apply(categorizar_color_exterior)
used_car_test['color_exterior_simple'] = used_car_test['exterior_color'].apply(categorizar_color_exterior)
used_car_train.drop(columns=["exterior_color"], axis=1, inplace=True)
used_car_test.drop(columns=["exterior_color"], axis=1, inplace=True)
# Verificar a distribuição
color_counts = used_car_train['color_exterior_simple'].value_counts()
print(color_counts)

color_exterior_simple
white     142350
silver    139571
black     126928
gray       66049
other      48717
red        36920
blue       34810
Name: count, dtype: int64


In [24]:
used_car_train.interior_color.value_counts().reset_index().head(60)

interior_color   count
0                     Black  226691
1                      Gray   39517
2                 Jet Black   36384
3                     Ebony   29999
4                  Charcoal   20765
5                     Beige   11937
6                  Graphite   11475
7               Titan Black    7408
8                       Tan    7292
9            Charcoal Black    4915
10                    Ivory    3350
11                      Red    3196
12      Diesel Gray / Black    3032
13                      Ash    2884
14                    Brown    2773
15                Parchment    2738
16               Light Gray    2102
17              Ebony Black    2022
18                    Wheat    1900
19        Medium Earth Gray    1899
20               Black Onyx    1845
21             Global Black    1789
22            Ebony / Ebony    1761
23          Dark Slate Gray    1757
24          Macchiato Beige    1553
25                   Cognac    1549
26       Medium Light Stone    1505
27                Dark Gray    1458
28                    White    1401
29         Black / Graphite    1387
30            Medium Pewter    1309
31          Dark Galvanized    1297
32                    Sport    1284
33  Black / Light Graystone    1264
34               Cappuccino    1238
35                Sandstone    1201
36                   Almond    1176
37              Black / Red    1141
38                    Blond    1090
39              Satin Black    1064
40                    Steel    1043
41       Silk Beige / Black    1040
42                    Mocha     974
43             Cocoa / Dune     967
44                    Cream     962
45                   Cement     954
46           Canberra Beige     946
47                    BLACK     943
48                 Cashmere     918
49            Light Neutral     900
50          Medium Ash Gray     877
51             Medium Stone     861
52                    Shale     813
53                    Taupe     810
54                   Pewter     776
55                    Cocoa     755
56                    Stone     739
57                     Dune     735
58                     Sand     715
59             Black / Gray     700

Como as cores acima são diferentes, logo terei que fazer uma função especifica para o interior do carro, a estratégia de agrupamento com base na amostra acima vai ser essa:

   * **Black / Jet Black / Ebony / Titan Black / Charcoal / Charcoal Black / Black Onyx / etc.** → agrupar como **Black**
   * **Gray / Light Gray / Medium Ash Gray / Dark Gray / Graphite / Steel / Shale / etc.** → agrupar como **Gray**
   * **Beige / Tan / Macchiato Beige / Sand / Sandstone / Almond / Cream / Parchment / etc.** → agrupar como **Beige**
   * **Red / Black / Red / Burgundy** → agrupar como **Red**
   * **White** → manter como **White**
   * **Outros raros** → agrupar como **Other**

In [25]:

def categorizar_interior_color(color):
    if pd.isna(color):
        return np.nan
    
    color_lower = str(color).lower()
    
    # Black
    if any(keyword in color_lower for keyword in ['black', 'ebony', 'jet black', 'onyx', 'charcoal', 'titan']):
        return 'black'
    
    # Gray
    if any(keyword in color_lower for keyword in ['gray', 'grey', 'graphite', 'steel', 'shale']):
        return 'gray'
    
    # Beige / Cream / Tan
    if any(keyword in color_lower for keyword in ['beige', 'tan', 'macchiato', 'sand', 'almond', 'cream', 'parchment', 'cappuccino', 'dune', 'cashmere', 'cognac']):
        return 'beige'
    
    # Red
    if 'red' in color_lower:
        return 'red'
    
    # White
    if 'white' in color_lower:
        return 'white'
    
    # Default para cores não mapeadas
    return 'other'


# Aplicar a categorização
used_car_train['color_interior_simple'] = used_car_train['interior_color'].apply(categorizar_interior_color)
used_car_test['color_interior_simple'] = used_car_test['interior_color'].apply(categorizar_interior_color)
used_car_train.drop(columns=["interior_color"], axis=1, inplace=True)
used_car_test.drop(columns=["interior_color"], axis=1, inplace=True)
# Verificar a distribuição
color_counts = used_car_train['color_interior_simple'].value_counts()
print(color_counts)

color_interior_simple
black    371829
gray      72095
other     58484
beige     45728
red        6840
white      2639
Name: count, dtype: int64


In [26]:
used_car_train.select_dtypes(include="object").nunique()

manufacturer              30
drivetrain                31
fuel_type                 35
mpg                      829
transmission_simple        3
color_exterior_simple      7
color_interior_simple      6
dtype: int64

A coluna manufacturer é responsável por informar a marca de cada veículo. Essa variável é importante, pois o fabricante influencia diretamente o preço, a percepção de valor e o perfil do carro.

In [27]:
used_car_train.manufacturer.value_counts().reset_index()

manufacturer  count
0            Ford  63062
1          Toyota  47420
2       Chevrolet  44473
3          Nissan  38297
4            Jeep  32988
5   Mercedes-Benz  32393
6           Honda  29932
7             BMW  29687
8             Kia  27540
9             GMC  23259
10          Dodge  19776
11         Subaru  19453
12     Volkswagen  19408
13        Hyundai  17609
14          Lexus  16719
15            RAM  15373
16           Audi  14165
17       Cadillac  13930
18          Mazda  12165
19          Buick  11452
20       Chrysler   9760
21       INFINITI   9737
22     Land Rover   9703
23        Porsche   9025
24        Lincoln   8433
25          Volvo   7958
26          Acura   6702
27          Tesla   4627
28     Mitsubishi   4584
29         Jaguar   2725

Como temos 30 valores unicos, irei pegar os 15 com maiores aparentes o resto irei deixar como others.

In [28]:
top_carros = used_car_train['manufacturer'].value_counts().nlargest(15).index
used_car_train['manufacturer'] = used_car_train['manufacturer'].apply(lambda x: x if x in top_carros else 'Other')
used_car_test['manufacturer'] = used_car_test['manufacturer'].apply(lambda x: x if x in top_carros else 'Other')




Agora irei analisar a coluna **`drivetrain`**, que indica o tipo de tração do carro. Essa coluna mostra se o veículo possui **tração dianteira, traseira, integral ou nas quatro rodas**, informações que podem influenciar no desempenho, preço e perfil de uso do carro. Vou observar os valores presentes para entender melhor a distribuição e identificar possíveis variações ou inconsistências que precisem ser tratadas.



In [29]:
used_car_train.drivetrain.value_counts().reset_index()

drivetrain   count
0                                   Front-wheel Drive  190902
1                                     All-wheel Drive  182233
2                                    Four-wheel Drive  124317
3                                    Rear-wheel Drive   77271
4                                                 FWD    4954
5                                                 AWD    2721
6                                                 4WD    1474
7                                                 RWD    1355
8                                     All-Wheel Drive      95
9                                             Unknown      81
10                                  Front-Wheel Drive      80
11  Four-Wheel Drive with Locking and Limited-Slip...      33
12   Front-Wheel Drive with Limited-Slip Differential      31
13  All-Wheel Drive with Locking and Limited-Slip ...      24
14                                   Four-Wheel Drive      24
15                                   Rear-Wheel Drive      22
16                                   Four Wheel Drive      16
17                                  Front Wheel Drive      14
18          All-Wheel Drive with Locking Differential      11
19    Rear-Wheel Drive with Limited-Slip Differential       9
20                                    All Wheel Drive       9
21                                   Rear Wheel Drive       8
22         Four-Wheel Drive with Locking Differential       5
23                                                4x4       5
24       Engine: 2.4L GDI DOHC 16V I4 -inc: dual CVVT       4
25                                                4x2       2
26                           Engine: 2.4L DOHC I4 GDI       1
27                                                2WD       1
28                                                4X4       1
29   Engine: 2.0L 4-Cylinder MPI 16V DOHC w/Dual CVVT       1
30     Engine: Lambda II 3.8L GDI DOHC V6 w/Dual CVVT       1

Percebe-se que tem muitas categorias com poucas amostras e que signifiam a mesma coisa, logo irei dividir assim:

1. **Definir categorias principais**
   A partir da amostra, podemos criar 4 categorias principais:

   * **FWD (Front-Wheel Drive / Tração dianteira)**
   * **RWD (Rear-Wheel Drive / Tração traseira)**
   * **AWD (All-Wheel Drive / Tração integral)**
   * **4WD (Four-Wheel Drive / 4x4 / Tração 4 rodas)**

2. **Mapear abreviações e variações de escrita**

   * Ex.:

     * “FWD”, “Front-wheel Drive”, “Front Wheel Drive” → **FWD**
     * “RWD”, “Rear-wheel Drive” → **RWD**
     * “AWD”, “All-wheel Drive”, “All Wheel Drive” → **AWD**
     * “4WD”, “Four-wheel Drive”, “4x4” → **4WD**

3. **Tratar valores raros ou não especificados**

   * Valores como “Unknown”, ou descrições longas tipo “Four-Wheel Drive with Locking and Limited-Slip…” podem ser agrupados na categoria correspondente se possível, ou em **Other/Unknown** se não for claro.


In [30]:

def categorizar_drivetrain(drivetrain):
    if pd.isna(drivetrain):
        return np.nan
    
    dt_lower = str(drivetrain).lower()
    
    # FWD - Front Wheel Drive
    if any(keyword in dt_lower for keyword in ['fwd', 'front-wheel', 'front wheel']):
        return 'FWD'
    
    # RWD - Rear Wheel Drive
    if any(keyword in dt_lower for keyword in ['rwd', 'rear-wheel', 'rear wheel']):
        return 'RWD'
    
    # AWD - All Wheel Drive
    if any(keyword in dt_lower for keyword in ['awd', 'all-wheel', 'all wheel']):
        return 'AWD'
    
    # 4WD - Four Wheel Drive / 4x4
    if any(keyword in dt_lower for keyword in ['4wd', 'four-wheel', 'four wheel', '4x4']):
        return '4WD'
    
    # Unknown / Other
    return 'Other'


# Aplicar a categorização
used_car_train['drivetrain_simple'] = used_car_train['drivetrain'].apply(categorizar_drivetrain)
used_car_test['drivetrain_simple'] = used_car_test['drivetrain'].apply(categorizar_drivetrain)
used_car_train.drop(columns=["drivetrain"], axis=1, inplace=True)
used_car_test.drop(columns=["drivetrain"], axis=1, inplace=True)
# Verificar a distribuição
color_counts = used_car_train['drivetrain_simple'].value_counts()
print(color_counts)

drivetrain_simple
FWD      195981
AWD      185093
4WD      125875
RWD       78665
Other        91
Name: count, dtype: int64


Agora irei analisar o tipo do combustivel do carro o fuel_type.

In [31]:
used_car_train.fuel_type.value_counts().reset_index()

fuel_type   count
0                        Gasoline  509848
1                          Hybrid   23169
2                          Diesel   22081
3                   E85 Flex Fuel   14820
4                        Electric   12876
5                               B    1126
6                   Flexible Fuel     371
7                  Plug-In Hybrid      97
8                   Gasoline Fuel      53
9   Gasoline/Mild Electric Hybrid      50
10               Regular Unleaded      45
11               Premium Unleaded      41
12                              G      38
13         Compressed Natural Gas      26
14                    Unspecified      19
15                    Hybrid Fuel       4
16                    Diesel Fuel       3
17                        Gaseous       3
18             Hydrogen Fuel Cell       2
19                          Other       2
20           Flex Fuel Capability       2
21           Electric Fuel System       2
22                           PHEV       2
23           Plug-In Electric/Gas       1
24                      Automatic       1
25                      Biodiesel       1
26                    Natural Gas       1
27                         E85 Fl       1
28                        Premium       1
29                      Flex Fuel       1
30            Gas/Electric Hybrid       1
31             Premium (Required)       1
32                     Bio Diesel       1
33                        Bi-Fuel       1
34                            Gas       1

Estratégia de Agrupamento para `fuel_type`

1. **Gasoline**

   * Inclui: `Gasoline`, `Gasoline Fuel`, `Regular Unleaded`, `Premium Unleaded`, `Flex Fuel Capability` (se fizer sentido)
   * Categoria final: **Gasoline**

2. **Hybrid**

   * Inclui: `Hybrid`, `Plug-In Hybrid`, `Gasoline/Mild Electric Hybrid`, `Hybrid Fuel`, `PHEV`, `Plug-In Electric/Gas`
   * Categoria final: **Hybrid**

3. **Diesel**

   * Inclui: `Diesel`, `Diesel Fuel`
   * Categoria final: **Diesel**

4. **Electric**

   * Inclui: `Electric`, `Electric Fuel System`, `Hydrogen Fuel Cell` (opcional, dependendo do objetivo)
   * Categoria final: **Electric**

5. **Outros / Não especificados**

   * Inclui: `E85 Flex Fuel`, `Flexible Fuel`, `Compressed Natural Gas`, `Unspecified`, `Other`, `B`, `F`, `G`
   * Categoria final: **Other** ou **Unknown**, se quiser tratar como missing.



In [32]:


def categorizar_fuel_type(fuel):
    if pd.isna(fuel):
        return np.nan
    
    fuel_lower = str(fuel).lower()
    
    # Gasoline
    if any(keyword in fuel_lower for keyword in ['gasoline', 'regular unleaded', 'premium unleaded', 'flex fuel capability']):
        return 'Gasoline'
    
    # Hybrid
    if any(keyword in fuel_lower for keyword in ['hybrid', 'plug-in', 'phev', 'mild electric']):
        return 'Hybrid'
    
    # Diesel
    if any(keyword in fuel_lower for keyword in ['diesel']):
        return 'Diesel'
    
    # Electric
    if any(keyword in fuel_lower for keyword in ['electric', 'hydrogen']):
        return 'Electric'
    
    # Outros / Não especificados
    return 'Other'


# Aplicar a categorização
used_car_train['fuel_type_simple'] = used_car_train['fuel_type'].apply(categorizar_fuel_type)
used_car_test['fuel_type_simple'] = used_car_test['fuel_type'].apply(categorizar_fuel_type)
used_car_train.drop(columns=["fuel_type"], axis=1, inplace=True)
used_car_test.drop(columns=["fuel_type"], axis=1, inplace=True)
# Verificar a distribuição
color_counts = used_car_train['fuel_type_simple'].value_counts()
print(color_counts)

fuel_type_simple
Gasoline    510039
Hybrid       23274
Diesel       22086
Other        16413
Electric     12880
Name: count, dtype: int64


A coluna mpg indica a eficiência de combustível do veículo, geralmente medida em milhas por galão. Essa variável é importante, pois impacta diretamente no custo de operação do carro e pode influenciar o preço e a preferência dos consumidores

In [33]:
used_car_train.mpg.value_counts().reset_index().head(50)

mpg  count
0   19-26  12783
1   18-25  12456
2   17-25  11280
3   20-27  10762
4   16-23   9131
5   19-28   8818
6   21-28   8324
7   17-23   8275
8   22-29   7965
9   17-24   7336
10  19-24   7280
11  16-22   7093
12  21-29   6880
13  21-27   6728
14  20-28   6598
15  20-26   6458
16  16-21   6136
17  24-30   6112
18  15-21   6096
19  22-30   6083
20  18-24   6059
21  19-27   6024
22  22-28   5833
23    0-0   5764
24  22-31   5657
25  19-25   5180
26  17-22   5150
27  26-33   4825
28  23-30   4811
29  25-32   4714
30  18-26   4573
31  18-27   4480
32  20-29   4433
33  24-31   4300
34  21-26   4165
35  21-30   4041
36  14-19   3974
37  23-29   3946
38  19-30   3933
39  23-34   3913
40  22-27   3880
41  18-23   3818
42  15-20   3776
43  25-30   3712
44  27-33   3660
45  26-31   3646
46  30-38   3606
47  13-17   3497
48  30-40   3444
49  28-39   3335

Os valores esão muito bagunçadosEstratégia para normalizar `mpg`

1. **Transformar intervalos em números únicos**

   * Para valores do tipo `X-Y`, você pode calcular a **média** `(X+Y)/2`.
   * Isso converte todos os intervalos em um único valor numérico, útil para modelagem.

2. **Tratar valores invertidos ou inválidos**

   * Se `X > Y`, inverter ou calcular a média normalmente `(X+Y)/2`.
   * Se houver zeros absurdos (como `13-0` ou `0-30`), tratar como **missing** ou ignorar, dependendo da frequência.

3. **Valores únicos / raros**

   * Entradas como `86-86` podem ser usadas normalmente (o intervalo vira `86`).
   * Valores isolados podem ser mantidos, mas você pode limitar o **range aceitável** para detectar outliers extremos (ex.: `mpg < 5` ou `mpg > 60`).

4. **Padronização final**

   * Converter tudo para **float**


Após isso eu irei ver seu valor preditivos, se não tiver muito valor preditivo eu a removo pelo falo que na imputação dos valores como ela tem 18% de valores vazios seria um risco de vies, logo irei fazer esse tratamento e analisar essas coisas.

In [34]:
import re
def tratar_mpg(mpg_value):
    """
    Função para normalizar valores da coluna mpg.
    - Converte intervalos X-Y em média.
    - Trata valores invertidos ou inválidos.
    - Converte tudo para float, mantendo NaN quando necessário.
    """
    if pd.isna(mpg_value):
        return np.nan
    
    mpg_str = str(mpg_value).strip()
    
    # Verifica se é um intervalo
    if '-' in mpg_str:
        parts = mpg_str.split('-')
        try:
            nums = [float(p) for p in parts if re.match(r'^\d+(\.\d+)?$', p)]
            if len(nums) == 2:
                # Se os números são válidos, calcula a média
                return np.mean(nums)
            elif len(nums) == 1:
                # Se só há um número válido, retorna ele
                return nums[0]
            else:
                return np.nan
        except:
            return np.nan
    else:
        # Tenta converter valor único para float
        try:
            return float(mpg_str)
        except:
            return np.nan

# Aplicar a categorização
used_car_train['mpg'] = used_car_train['mpg'].apply(tratar_mpg)
used_car_test['mpg'] = used_car_train['mpg'].apply(tratar_mpg)

used_car_train[['mpg', 'price']].corr()

mpg     price
mpg    1.000000 -0.244032
price -0.244032  1.000000

Decidi **manter a coluna `mpg`** no dataset, pois ela apresenta uma correlação negativa relevante com o preço (`-0.244`), indicando que veículos mais econômicos tendem a ter preços ligeiramente menores. Apesar de não ser extremamente forte, essa relação ainda fornece informação útil que pode contribuir para o desempenho do modelo.

Para lidar com os valores ausentes, irei **aplicar uma imputação**, garantindo que todos os registros tenham um valor numérico consistente. Com isso, a coluna estará pronta para ser utilizada na modelagem, preservando sua capacidade preditiva.
Com a parte de limpeza finalizada, posso salvar os 2 data sets para utiliza-los na analise exploratória!


In [ ]:
used_car_test.to_csv("../Dados/used_car_test.csv")
used_car_train.to_csv("../Dados/used_car_train.csv")